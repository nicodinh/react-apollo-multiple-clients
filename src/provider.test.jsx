import React, { Component } from 'react';
import renderer from 'react-test-renderer';
import { objectOf, any } from 'prop-types';
import { Query as ApolloQuery, Mutation as ApolloMutation } from 'react-apollo';

import { ApolloMultipleClientsProvider, withClient } from '.';

/* eslint-disable react/prop-types */
jest.mock('react-apollo', () => ({
  Query: ({ client, ...rest }) => <div className="query" {...rest}>{client}</div>,
  Mutation: ({ client, ...rest }) => <div className="mutation" {...rest}>{client}</div>,
}));
/* eslint-enable react/prop-types */

const client1 = 'client1';
const client2 = 'client2';

const clientList = {
  clientName1: client1,
  clientName2: client2,
};

describe('React Apollo Multiple Clients', () => {
  /* eslint-disable no-console */
  beforeAll(() => {
    global.consoleError = console.error;
    console.error = () => {};
  });
  afterAll(() => {
    console.error = global.consoleError;
  });

  it('should throw error when client cannot be found', () => {
    try {
      const Enriched = withClient()(ApolloMutation);
      renderer.create(
        <ApolloMultipleClientsProvider clients={clientList}>
          <Enriched />
        </ApolloMultipleClientsProvider>,
      );
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it('should provide correct client', () => {
    const EnrichedQuery = withClient('test')(ApolloQuery);
    const EnrichedMutation = withClient('test')(ApolloMutation);

    const app = renderer.create(
      <ApolloMultipleClientsProvider clients={{ test: 'clientos' }}>
        <EnrichedQuery />
        <EnrichedMutation />
      </ApolloMultipleClientsProvider>,
    );

    expect(app.toJSON()).toMatchSnapshot();
  });

  it('should not override client outside provider', () => {
    const Enriched = withClient('test')(ApolloQuery);

    const app = renderer.create(
      <div>
        <Enriched client="foobar-NOT-overridden" />
      </div>,
    );

    expect(app.toJSON()).toMatchSnapshot();
  });

  it('should override client inside provider', () => {
    const Enriched = withClient('test')(ApolloMutation);

    const app = renderer.create(
      <div>
        <ApolloMultipleClientsProvider clients={{ test: 'clientos' }}>
          <Enriched client="foobar-overridden" />
        </ApolloMultipleClientsProvider>
      </div>,
    );

    expect(app.toJSON()).toMatchSnapshot();
  });

  it('should work with class as well', () => {
    class Foo extends Component {
      constructor(props) {
        super(props);
        this.state = {};
      }

      render() {
        const { client } = this.props;
        return <div>{client}</div>;
      }
    }
    Foo.propTypes = {
      client: objectOf(any).isRequired,
    };

    const Enriched = withClient('test')(Foo);

    const app = renderer.create(
      <ApolloMultipleClientsProvider clients={{ test: 'clientos' }}>
        <Enriched />
      </ApolloMultipleClientsProvider>,
    );

    expect(app.toJSON()).toMatchSnapshot();
  });
});
