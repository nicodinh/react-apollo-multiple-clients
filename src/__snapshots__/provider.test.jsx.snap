// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Apollo Multiple Clients should not override client outside provider 1`] = `
<div>
  <div
    className="provider null"
  >
    <div
      className="query"
      client="foobar-NOT-overridden"
    >
      <span />
    </div>
  </div>
</div>
`;

exports[`React Apollo Multiple Clients should override client inside provider 1`] = `
<div>
  <div
    className="provider clientos"
  >
    <div
      className="query"
      client="foobar-overridden"
    >
      <span>
        clientos
      </span>
    </div>
  </div>
</div>
`;

exports[`React Apollo Multiple Clients should provide correct client 1`] = `
Array [
  <div
    className="provider client1"
  >
    <div
      className="query"
    >
      <span>
        client1
      </span>
    </div>
  </div>,
  <div
    className="provider client2"
  >
    <div
      className="query"
    >
      <span>
        client2
      </span>
    </div>
  </div>,
]
`;

exports[`React Apollo Multiple Clients should provide correct client within nested provider 1`] = `
Array [
  <div
    className="provider client1"
  >
    <div
      className="query"
    >
      <span>
        client1
      </span>
    </div>
  </div>,
  <div
    className="provider client1"
  >
    <div
      className="query"
    >
      <span>
        client1
      </span>
      <div
        className="query"
      >
        <span>
          client1
        </span>
      </div>
      <div
        className="provider client2"
      >
        <div
          className="query"
        >
          <span>
            client2
          </span>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="provider client2"
  >
    <div
      className="query"
    >
      <span>
        client2
      </span>
    </div>
  </div>,
]
`;

exports[`React Apollo Multiple Clients should throw error when client cannot be found 1`] = `[Error: Could not find client not-available]`;

exports[`React Apollo Multiple Clients should throw error when client no client is given 1`] = `[Error: Please provide a client namespace]`;

exports[`React Apollo Multiple Clients should throw error when provider has no clients 1`] = `[Error: You need to define at least one client]`;

exports[`React Apollo Multiple Clients should work with class as well 1`] = `
<div
  className="provider clientos"
>
  <div />
</div>
`;
